### ISSUE TITLE
케라스와 신경망 : 훑어보기(Tensor Flow 2.0)

### ISSUE THREAD

**Issue by iju1633** [iju1633/Fintech_Flip_Learning_contents#3]
### 사람이 규칙을 정하기 어려운 경우에 사용 
- 블랙 박스 안에서 무슨 일이 일어나는 지 정확히 사람이 정의하기 어려운 경우
### 정확도가 높으면 좋은 모델일까?
- 코로나 진단 키드 모델
- 100명 한테 했더니 정확도가 99%인 진단키트입니다.
- 실제로 100명이 코로나 음성인 사람이 99명이었고, 1명이 코로나 양성이었는데,
- 진단 키트가 전부 다 음성이라고 예측한 것이었다.
- 정확도를 99%라고 말할 수 있지만,,양성인데 음성이라고 판단한 사례가 있는 것이다.
### 신용카드 사기 탐지
- 정상 거래가 98건, 사기 거래가 2건
- 모델이 전부 정상 거래로 예측한다면, 정확도는 98%이지만, 사기임에도 정상거래라고 판단하는 경우가 있다..
### 위의 2가지 경우를, 레이블이 불균형한 경우라고 한다
- 이때 정확도는 올바른 측정 방법이 아닐 수 있다
- 그러므로 이런 불균형을 반영한 F1-score라는 것을 사용합니다.
- ### Precision, Recall, Accuracy, F1-score 모두 알고 있을 것

## 딥러닝
- 머신러닝의 종류로, 인공 신경망(알고리즘)을 사용하는데, 다른 머신러닝의 알고리즘보다 성능이 좋은 경우가 대부분이다
### 데이터를 가장 잘 반영하는 직선을 긋는 것 -> 예측 용이
- y = wx + b ( w : 가중치, b : 편향)
- Mean Square error가 가장 작도록 하는 w와 b(매개변수)를 찾는 것이 선형 회귀의 목표, 딥러닝도 비슷하게 손실 함수가 최소로 되도록 하는 w와 b를 찾는 것이 목표

### cost function이 최소화 되게끔 하는 알고리즘을 설계하는 것이 남은 과제
- Optimizer ex) Gradient Descent(경사 하강법) : 다음 W값 = 현재의 W값 – 접선에서의 기울기
- Learning Rate : 0에서 1 사이, 극단적으로 하강하는 경우를 없애주기 위함
    - 0에서 1사이 값을 주는데, 아무거나 넣으면 안됨 -> 적당한 값을 찾아야함 -> 큰 값에서 반씩 줄여가며 or 높은 성능을 얻었던 논문에서 사용하는 Learning rate 값 사용

### 두 개의 선택지에서 분별하는 문제는 하나는 1, 하나는 0으로 정의한다는 것을 알고 있자
- sigmoid 함수를 통해 분류 알고리즘을 짤 수 있다, 그림 형태 기억하고 이름을 기억할 것
- 출력값이 0.5 이상이면 1(True), 0.5미만이면 0(False)로 만들면 **이진 분류 문제**로 사용
- 로지스틱 회귀는 비용 함수로 평균 제곱 오차를 사용하지 않는다.
    - 비용 함수로 평균 제곱 오차를 사용할 경우, 바로 local minimum으로 빠지기에 최적인 w와 b를 구할 수 없음
    - 비용함수로 크로스 엔트로피 함수 사용 : **최대한 밥그릇 모양 비슷하게**  함수를 만들어서 경사 하강법을 사용하기 위함
- code에서 activation은 wx+b에 씌울 함수를 의미한다
    - activation = 'linear' : 아무것도 씌우지 않겠다 : wx+b에 **sigmoid()와 같은 걸 씌우지 않겠다는 의미**
    - activation = 'sigmoid' : 시그모이드를 씌우겠다
- 딥러닝 그림에서 화살표는 w(가중치)를 의미한다
- 주어진 데이터에 따라 최적의 은닉층의 개수, 동그라미 수는 정해져 있지 않다
### 어디로 분류될 지 알아서 파악해주는 것
- 딥러닝은 텍스트를 숫자로 바꿔서 넣어주어야 함(vectorization)
### Word Embedding하면서 RNN을 하는 경우에 대해 알아볼 예정(여태 Dense 써왔음)
- 사람처럼 단어를 순차적으로 읽음, 입력할 때 단어 하나 씩 입력함
- RNN은 **과거의 정보를 기억**하고 있다(식을 보게 되면 전단계가 매개변수에 포함되어 있음을 알 수 있음)
- 하이퍼볼릭탄젠트 함수는 시그모이드 함수와 달리 -1 ~ 1의 범위 값으로 값을 반환하는 함수이다
    - 반환값의 범위가 시그모이드 함수보다 크므로 일반적으로 은닉층에서는 시그모이드 함수보다 더 잘 동작한다
- 기존의 RNN은 시점이 길어지면서 앞에 있던 정보가 소실되는 ‘**장기 의존성 문제**’를 갖고 있다
    - RNN의 고질적인 문제(점점 색이 옅어짐)
    - 이에 따라 LSTM과 GRU를 사용(더욱 복잡한 모델)
- 순차적으로 데이터를 읽기 때문에 주가 예측에 용이하다.
